hidimstat/CPI.py

```diff
@@ -0,0 +1,180 @@
+            self.groups = {j: [j] for j in range(self.nb_groups)}
+        else:
+            self.nb_groups = len(self.groups)
+        # create a list of covariate estimators for each group if not provided
+        if len(self.list_cov_estimators) == 0:
+            self.list_cov_estimators = [clone(self.covariate_estimator)
+                                        for _ in range(self.nb_groups)]
+
+        def joblib_fit_one_gp(estimator, X, y, j):
+            """
+            Fit a single covariate estimator to predict a single group of
+            covariates.
+            """
+            X_j = X[:, self.groups[j]].copy()
+            X_minus_j = np.delete(X, self.groups[j], axis=1)
+            estimator.fit(X_minus_j, X_j)
+            return estimator
+
+        # Parallelize the fitting of the covariate estimators
+        self.list_cov_estimators = Parallel(n_jobs=self.n_jobs)(
+            delayed(joblib_fit_one_gp)(estimator, X, y, j)
+            for j, estimator in enumerate(self.list_cov_estimators))
+
+        return self
+
+    def predict(self, X, y):
+        """
+        Compute the CPI importance scores. For each group of covariates, the
+        residuals are computed using the covariate estimators. The residuals
+        are then permuted and the model is re-evaluated. The importance score
+        is the difference between the loss of the model with the original data
+        and the loss of the model with the permuted data.
+
+        Parameters
+        ----------
+        X: array-like of shape (n_samples, n_features)
+            The input samples.
+        y: array-like of shape (n_samples,)
+            The target values.
+
+        Returns
+        -------
+        output_dict: dict
+            A dictionary containing the following keys:
+            - 'loss_reference': the loss of the model with the original data.
+            - 'loss_perm': a dictionary containing the loss of the model with
+            the permuted data for each group.
+            - 'importance': the importance scores for each group.
+        """
+        if len(self.list_cov_estimators) == 0:
+            raise ValueError("fit must be called before predict")
+        for m in self.list_cov_estimators:
+            check_is_fitted(m)
+
+        output_dict = dict()
+        if self.score_proba:
+            y_pred = self.estimator.predict_proba(X)
+        else:
+            y_pred = self.estimator.predict(X)
+        loss_reference = self.loss(y_true=y, y_pred=y_pred)
+        output_dict["loss_reference"] = loss_reference
+        output_dict['loss_perm'] = dict()
+
+        def joblib_predict_one_gp(estimator, X, y, j):
+            """
+            Compute the importance score for a single group of covariates.
+            """
+            list_loss_perm = []
+            X_j = X[:, self.groups[j]].copy()
+            X_minus_j = np.delete(X, self.groups[j], axis=1)
+            X_j_hat = estimator.predict(X_minus_j).reshape(X_j.shape)
```

Comment:
So there is no cross-fitting on this ? I think that this may degrade the importance accuracy.